src/tensor/visitors/base.rs

```diff
@@ -0,0 +1,243 @@
+        &mut self,
+        get_refs: GetRef,
+        get_muts: GetMut,
+        opts: TensorOptions<S, E, D>,
+    ) -> Result<(), Self::Err>
+    where
+        GetRef: FnMut(&T) -> &Tensor<S, E, D>,
+        GetMut: FnMut(&mut T) -> &mut Tensor<S, E, D>;
+}
+
+pub(crate) struct RecursiveWalker<'a, M, F> {
+    pub(crate) m: M,
+    pub(crate) f: &'a mut F,
+    pub(crate) path: &'a mut Vec<String>,
+}
+
+impl<'a, M, E: Dtype, D: DeviceStorage, F: VisitTensorRef<E, D>> ModuleWalker<M, E, D>
+    for RecursiveWalker<'a, &'a M, F>
+{
+    type Err = F::Err;
+    fn visit_module<Field, GetRef, GetMut>(
+        &mut self,
+        mut get_refs: GetRef,
+        _: GetMut,
+        name: &str,
+    ) -> Result<(), Self::Err>
+    where
+        GetRef: FnMut(&M) -> &Field,
+        GetMut: FnMut(&mut M) -> &mut Field,
+        Field: TensorCollection<E, D>,
+    {
+        self.path.push(name.into());
+        let mut walker = RecursiveWalker {
+            m: get_refs(self.m),
+            f: self.f,
+            path: self.path,
+        };
+        Field::iter_tensors(&mut walker)?;
+        self.path.pop();
+        Ok(())
+    }
+    fn visit_tensor<S: Shape, GetRef, GetMut>(
+        &mut self,
+        mut get_refs: GetRef,
+        _: GetMut,
+        opts: TensorOptions<S, E, D>,
+    ) -> Result<(), F::Err>
+    where
+        GetRef: FnMut(&M) -> &Tensor<S, E, D>,
+        GetMut: FnMut(&mut M) -> &mut Tensor<S, E, D>,
+    {
+        self.path.push(opts.name.into());
+        self.f.visit(self.path.join("."), opts, get_refs(self.m))?;
+        self.path.pop();
+        Ok(())
+    }
+}
+
+impl<'a, M, E: Dtype, D: DeviceStorage, F: VisitTensorMut<E, D>> ModuleWalker<M, E, D>
+    for RecursiveWalker<'a, &'a mut M, F>
+{
+    type Err = F::Err;
+    fn visit_module<Field, GetRef, GetMut>(
+        &mut self,
+        _: GetRef,
+        mut get_muts: GetMut,
+        name: &str,
+    ) -> Result<(), F::Err>
+    where
+        GetRef: FnMut(&M) -> &Field,
+        GetMut: FnMut(&mut M) -> &mut Field,
+        Field: TensorCollection<E, D>,
+    {
+        self.path.push(name.into());
+        let mut walker = RecursiveWalker {
+            m: get_muts(self.m),
+            f: self.f,
+            path: self.path,
+        };
+        Field::iter_tensors(&mut walker)?;
+        self.path.pop();
+        Ok(())
+    }
+    fn visit_tensor<S: Shape, GetRef, GetMut>(
+        &mut self,
+        _: GetRef,
+        mut get_muts: GetMut,
+        opts: TensorOptions<S, E, D>,
+    ) -> Result<(), F::Err>
+    where
+        GetRef: FnMut(&M) -> &Tensor<S, E, D>,
+        GetMut: FnMut(&mut M) -> &mut Tensor<S, E, D>,
+    {
+        self.path.push(opts.name.into());
+        self.f.visit(self.path.join("."), opts, get_muts(self.m))?;
+        self.path.pop();
+        Ok(())
+    }
+}
+
+impl<'a, M, E: Dtype, D: DeviceStorage, F: VisitTensorMutRef<E, D>> ModuleWalker<M, E, D>
+    for RecursiveWalker<'a, (&'a mut M, &'a M), F>
```

Comment:
This impl is what enables EMA. Basically we'd just need to impl VisitTensorMutRef for some struct Ema and add an extension method